### PR TITLE
Add module xgrammar.contrib.mlxlm

### PR DESCRIPTION
**Pull request by wangkuiyi**
After a chat with @Ubospica , I am filing this PR to fix https://github.com/mlc-ai/xgrammar/issues/270

This module provides a logit processor for mlx_lm, just like hf.py provides one for Hugging Face Transformers.

This module is runnable and works demonstrating that XGrammar runs on Apple Silicon via MLX and mlx-lm, just like it works on CUDA via PyTorch and Hugging Face Transformers.

You can run this module like the following

```shell
python -m mlxlm \
 --model mlx-community/Qwen2.5-Coder-3B-Instruct-4bit \
 --prompt "Generate a very simple example JSON. No text. Only the JSON."
```

It generates a grammar-verified JSON like the following:

```json
{
  "name": "John Doe",
  "age": 30,
  "city": "New York"
}
```

### Optional Dependency

This pull request adds mlx-lm as an optional dependency `apple-silicon`. To run this demo modules, you need to install XGrammar from source code:

```shell
pip install -e .[apple-silicon]
```

### Next Steps

1. Currently, this demo module expand the bitmask generated from the pushdown automaton of XGrammar into a float array before adding it to the vocabulary logits generated by the LLM. We should add a Apple Silicon GPU kernel for MLX, just like the CUDA, Triton and CPU kernels for PyTorch.  In this way, we don't expand the bitmask and saves application time.

2. Currently, XGrammar generates the bitmask as a PyTorch tensor and this demo module converts it into an MLX array. I haven't benchmarked the cost of the conversion. If it is high, we may consider enabling XGrammar generating the bitmask as an MLX array optionally.

### CODE REVIEW

**Review by Ubospica — APPROVED** [review]
LGTM. Thank you so much for your contribution! @wangkuiyi
